libnmstate/nm/bridge_port_vlan.py

```diff
@@ -0,0 +1,221 @@
+
+    @property
+    def is_native_vlan_enabled(self):
+        return self._is_native
+
+    @property
+    def port_type(self):
+        return self._port_mode
+
+    def to_nm(self):
+        """
+        Generate a list of NM.BridgeVlan objects from the encapsulated
+        PortVlanFilter data
+        """
+
+        port_vlan_config = []
+        if self._port_mode == LB.Port.Vlan.Mode.TRUNK:
+            port_vlan_config += map(
+                PortVlanFilter._generate_vlan_trunk_port_config,
+                self._trunk_tags,
+            )
+            if self.is_native_vlan_enabled and self._tag:
+                port_vlan_config.append(
+                    PortVlanFilter._generate_vlan_access_port_config(self._tag)
+                )
+        elif self._port_mode == LB.Port.Vlan.Mode.ACCESS and self._tag:
+            port_vlan_config.append(
+                PortVlanFilter._generate_vlan_access_port_config(self._tag)
+            )
+
+        return port_vlan_config
+
+    def to_dict(self):
+        """
+        Get the port vlan filtering configuration in dict format - e.g. in yaml
+        format:
+        - name: eth1
+          vlan:
+            type: trunk
+            trunk-tags:
+              - id: 101
+              - id-range:
+                  min: 200
+                  max: 4095
+            tag: 100
+            enable-native: true
+        """
+
+        port_vlan_state = {
+            LB.Port.Vlan.MODE: self.port_type,
+            LB.Port.Vlan.TRUNK_TAGS: self.trunk_ports,
+        }
+        if self.tag:
+            port_vlan_state[LB.Port.Vlan.TAG] = self.tag
+        if self.port_type == LB.Port.Vlan.Mode.TRUNK:
+            port_vlan_state[
+                LB.Port.Vlan.ENABLE_NATIVE
+            ] = self.is_native_vlan_enabled
+        return port_vlan_state
+
+    def import_from_bridge_settings(self, nm_bridge_vlans):
+        """
+        Instantiates a PortVlanFilter object from a list of NM.BridgeVlan
+        objects.
+        """
+
+        self._is_native = False
+        trunk_tags = []
+
+        for nm_bridge_vlan in nm_bridge_vlans:
+            vlan_min, vlan_max = PortVlanFilter.get_vlan_tag_range(
+                nm_bridge_vlan
+            )
+            if PortVlanFilter._is_access_port(nm_bridge_vlans):
+                self._tag = vlan_min
+            elif nm_bridge_vlan.is_pvid() and nm_bridge_vlan.is_untagged():
+                # an NM.BridgeVlan has a range and can be PVID and/or untagged
+                # according to NM's model, PVID / untagged apply to the 'max'
+                # part of the range
+                self._tag = vlan_max
+                self._is_native = True
+            else:
+                trunk_tags.append(
+                    PortVlanFilter._translate_nm_bridge_vlan_to_trunk_tags(
+                        vlan_min,
+                        vlan_max,
+                        nm_bridge_vlan.is_pvid(),
+                        nm_bridge_vlan.is_untagged(),
```

Comment:
You need to include the native vlan in the "trunk tags"? AFAIU the PVID and UNTAGGED are not coming from the nmstate desired state, so it probably should not be included here as well.
I was under the impression that these internal constants were needed only for the intermediate structure, which are no longer needed now.
But maybe I miss something, please clarify.

Reply:
Not needed, since the native vlan goes into a different flow (set `is_native` & `tag`).

For trunk tags, all we need is the range. Good catch.